crypto4pl.c

```diff
@@ -66,6 +68,255 @@ typedef enum
 } crypt_mode_t;
 
 
+                 /***************************
+                 *         HASHING          *
+                 ****************************/
+
+
+#define CONTEXT_MAGIC (~ 0x53481284L)
+
+typedef struct context
+{ int             magic;
+  atom_t          atom;
+  unsigned int    encoding;
+  const EVP_MD   *algorithm;
+  EVP_MD_CTX     *ctx;
+} PL_CRYPTO_CONTEXT;
+
+
+static int
+release_context(atom_t atom)
+{ PL_CRYPTO_CONTEXT *c;
+  size_t size;
+
+  c = PL_blob_data(atom, &size, NULL);
+  ssl_deb(4, "Releasing PL_CRYPTO_CONTEXT %p\n", c);
+  EVP_MD_CTX_free(c->ctx);
+  free(c);
+  return TRUE;
+}
+
+static int
+compare_context(atom_t a, atom_t b)
+{ PL_CRYPTO_CONTEXT* *c1 = PL_blob_data(a, NULL, NULL);
+  PL_CRYPTO_CONTEXT* *c2 = PL_blob_data(b, NULL, NULL);
+
+  return ( c1 > c2 ?  1 :
+           c1 < c2 ? -1 : 0
+         );
+}
+
+static int
+write_context(IOSTREAM *s, atom_t symbol, int flags)
+{ PL_CRYPTO_CONTEXT *c = PL_blob_data(symbol, NULL, NULL);
+
+  Sfprintf(s, "<crypto_context>(%p)", c);
+
+  return TRUE;
+}
+
+static void
+acquire_context(atom_t atom)
+{ ssl_deb(4, "Acquire on atom %d\n", atom);
+}
+
+static PL_blob_t crypto_context_type =
+{ PL_BLOB_MAGIC,
+  PL_BLOB_NOCOPY,
+  "crypto_context",
+  release_context,
+  compare_context,
+  write_context,
+  acquire_context
+};
+
+static int
+put_context(term_t tcontext, PL_CRYPTO_CONTEXT *context)
+{ return PL_unify_atom(tcontext, context->atom);
+}
+
+static int
+register_context(term_t tcontext, PL_CRYPTO_CONTEXT *context)
+{ term_t blob = PL_new_term_ref();
+  int rc;
+
+  PL_put_blob(blob, context, sizeof(void*), &crypto_context_type);
+  rc = PL_get_atom(blob, &context->atom);
+  assert(rc);
+  ssl_deb(4, "Atom created: %d\n", context->atom);
+  return put_context(tcontext, context);
+}
+
+static int
+get_context(term_t tcontext, PL_CRYPTO_CONTEXT **context)
+{ PL_blob_t *type;
+  void *data;
+
+  if ( PL_get_blob(tcontext, &data, NULL, &type) &&
+       type == &crypto_context_type )
+  { PL_CRYPTO_CONTEXT *c = data;
+
+    assert(c->magic == CONTEXT_MAGIC);
+    *context = c;
+
+    return TRUE;
+  }
+
+  return PL_type_error("crypto_context", tcontext);
+}
+
+
+
+static int
+hash_options(term_t options, PL_CRYPTO_CONTEXT *result)
+{ term_t opts = PL_copy_term_ref(options);
+  term_t opt = PL_new_term_ref();
+
+  /* defaults */
+  result->encoding = REP_UTF8;
+  result->algorithm = EVP_sha256();
+
+  while(PL_get_list(opts, opt, opts))
```

Comment:
If you use PL_get_list_ex and PL_get_nil_ex and simply return FALSE if they fail.